lib/exercises.rb

```diff
@@ -2,19 +2,62 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(w * n^2) ?
+# THINKING....
+# Iterating through the string = O(n), summing each word = O(w); O(n * w)
+# checking for the word's sum in the hash and shovel onto array = O(1)
+# returning the hash values = O(n) b/c if no words are anagrams, our hash would be the
+# same size as the original array.
+
+# Space Complexity: O(n); The hash ends up being the same size as the string/input size.
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  return strings if strings.empty?
+  
+  anagrams = {}
+  # for each word, if the sum is in the hash
+  strings.each do |word|
+    if anagrams[word.sum].nil?
+      anagrams[word.sum] = [word]
+    else
+      anagrams[word.sum] << word
+    end
+  end
+  
+  return anagrams.values
 end
 
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n + k)
+# Iterate through each element in order to count and add to hash O(n)
+# Sort the hash: maybe O(log n)
+# Put return first k elements = O(k)
+
+# Space Complexity: O(n); worst case, hash is same size as array. Other variables are O(1).
 def top_k_frequent_elements(list, k)
```

Comment:
I would say your time complexity is O(k + n log n), but since k should be less than n, this is O(n log n)